spec-tests/src/test/java/org/candlepin/spec/EntitlementCertificateSpecTest.java

```diff
@@ -0,0 +1,474 @@
+        List<CertificateDTO> oldCerts = consumerApi.fetchCertificates(system.getUuid());
+        assertThat(oldCerts).hasSize(2);
+        AsyncJobStatusDTO result = entitlementsApi.regenerateEntitlementCertificatesForProduct(
+            coolApp.getId(), false);
+        if (result != null) {
+            client.jobs().waitForJob(result.getId());
+        }
+
+        List<CertificateDTO> newCerts = consumerApi.fetchCertificates(system.getUuid());
+        assertEquals(oldCerts.size(), newCerts.size());
+        // System has two certs, but we only regenerated for one product, so the
+        // other serial should have remained the same:
+        assertThat(newCerts.stream()
+            .filter(element -> oldCerts.contains(element))
+            .collect(Collectors.toList())).hasSize(1);
+    }
+
+    @Test
+    @DisplayName("will be regenerated when changing existing subscription's end date")
+    public void regenerateOnChangeInSubEnd() throws Exception {
+        CertificateDTO oldCert = consumerApi.fetchCertificates(system.getUuid()).get(0);
+        if (CandlepinMode.isHosted()) {
+            subscription.setEndDate(subscription.getEndDate().plusDays(2L));
+            subscription = hostedTestApi.updateSubscription(subscription.getId(), subscription);
+            AsyncJobStatusDTO refresh = ownerApi.refreshPools(owner.getKey(), false);
+            if (refresh != null) {
+                client.jobs().waitForJob(refresh.getId());
+            }
+        }
+        else {
+            pool.setEndDate(pool.getEndDate().plusDays(2L));
+            ownerApi.updatePool(owner.getKey(), pool);
+        }
+        CertificateDTO newCert = consumerApi.fetchCertificates(system.getUuid()).get(0);
+        assertNotEquals(oldCert.getSerial().getId(), newCert.getSerial().getId());
+        EntitlementDTO ent = consumerApi.listEntitlements(system.getUuid()).get(0);
+        OffsetDateTime test = subscription != null ? subscription.getEndDate() : pool.getEndDate();
+        assertEquals(ent.getEndDate(), test);
+    }
+
+    @Test
+    @DisplayName("single entitlement in excess will be deleted when " +
+        "existing subscription quantity is decreased")
+    public void deleteEntWhenSubQuantityDecreased() throws Exception {
+        //  this entitlement makes the counts inconclusive
+        consumerApi.unbindAll(system.getUuid());
+        ProductDTO prod = Products.randomEng().addAttributesItem(
+            new AttributeDTO().name("multi-entitlement").value("yes"));
+        if (CandlepinMode.isHosted()) {
+            prod = hostedTestApi.createProduct(prod);
+            subscription = Subscriptions.random(owner, prod);
+            subscription = hostedTestApi.createSubscription(subscription);
+            AsyncJobStatusDTO refresh = ownerApi.refreshPools(owner.getKey(), false);
+            if (refresh != null) {
+                client.jobs().waitForJob(refresh.getId());
```

Comment:
We should validate that the job successfully completed. It looks like this will need to be done through out where the `waitForJob` method is being used.